modules/lib/lib-auth/src/main/resources/lib/xp/auth.ts

```diff
@@ -0,0 +1,832 @@
+    const bean = __.newBean<FindPrincipalsHandler>('com.enonic.xp.lib.auth.FindPrincipalsHandler');
+
+    const {
+        type,
+        idProvider,
+        start = 0,
+        count = 10,
+        name,
+        searchText,
+    } = params ?? {};
+
+    bean.setType(__.nullOrValue(type));
+    bean.setIdProvider(__.nullOrValue(idProvider));
+    bean.setStart(__.nullOrValue(start));
+    bean.setCount(__.nullOrValue(count));
+    bean.setName(__.nullOrValue(name));
+    bean.setSearchText(__.nullOrValue(searchText));
+
+    return __.toNativeObject(bean.findPrincipals());
+}
+
+interface DeletePrincipalHandler {
+    setPrincipalKey(value?: string | null): void;
+
+    deletePrincipal(): boolean;
+}
+
+/**
+ * Deletes the principal with the specified key.
+ *
+ * @example-ref examples/auth/deletePrincipal.js
+ *
+ * @param {string} principalKey Principal key to delete.
+ * @returns {boolean} True if deleted, false otherwise.
+ */
+export function deletePrincipal(principalKey: string): boolean {
+    const bean = __.newBean<DeletePrincipalHandler>('com.enonic.xp.lib.auth.DeletePrincipalHandler');
+    bean.setPrincipalKey(__.nullOrValue(principalKey));
+    return __.toNativeObject(bean.deletePrincipal());
+}
+
+interface GetIdProviderConfigHandler {
+    execute(): Record<string, Object>;
+}
+
+/**
+ * This function returns the ID provider configuration.
+ * It is meant to be called from an ID provider controller.
+ *
+ * @example-ref examples/auth/getIdProviderConfig.js
+ *
+ * @returns {object} The ID provider configuration as JSON.
+ */
+export function getIdProviderConfig(): Record<string, Object> {
+    const bean = __.newBean<GetIdProviderConfigHandler>('com.enonic.xp.lib.auth.GetIdProviderConfigHandler');
+    return __.toNativeObject(bean.execute());
+}
+
+export interface GetProfileParams {
+    key: string;
+    scope?: string;
+}
+
+interface GetProfileHandler {
+    setKey(value: string): void;
+
+    setScope(value?: string | null): void;
+
+    execute(): Record<string, Object>;
+}
+
+/**
+ * This function retrieves the profile of a user.
+ *
+ * @example-ref examples/auth/getProfile.js
+ *
+ * @param {object} params JSON parameters.
+ * @param {string} params.key Principal key of the user.
+ * @param {string} [params.scope] Scope of the data to retrieve.
+ * @returns {object} The extra data as JSON
+ */
+export function getProfile(params: GetProfileParams): Record<string, Object> {
+    required(params, 'key');
+
+    const bean = __.newBean<GetProfileHandler>('com.enonic.xp.lib.auth.GetProfileHandler');
+
+    bean.setKey(params.key);
+    bean.setScope(__.nullOrValue(params.scope));
+
+    return __.toNativeObject(bean.execute());
+}
+
+export interface ModifyProfileParams {
+    key: string;
+    scope?: string | null;
+    editor: (group: Group) => Group;
+}
+
+interface ModifyProfileHandler {
+    setKey(value: string): void;
+
+    setScope(value?: string | null): void;
+
+    setEditor(value: Object): void;
```

Comment:
All editors are not `Object`. Actually, it must be a Function, since we use `editor.call(/*...*/)` in Java.
Use `Function` for this purpose.
Or we can go even further and define an editor type on top of the file:
```ts
export type EditorFn = <T>(value: T) => T
```
Later we can do something like this:
```ts
editor: EditorFn<Group>;
// ...
setEditor(value: EditorFn<Group>): void;
```